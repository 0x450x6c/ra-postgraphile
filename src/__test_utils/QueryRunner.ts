/* eslint-disable @typescript-eslint/no-var-requires */
import { SchemaLink } from 'apollo-link-schema'
import { InMemoryCache } from 'apollo-cache-inmemory'
import { ApolloClient } from 'apollo-client'
require('dotenv').config()
import { Pool } from 'pg'
import { createPostGraphileSchema, withPostGraphileContext } from 'postgraphile'

import PgSimplifyInflectorPlugin from '@graphile-contrib/pg-simplify-inflector'
// @ts-ignore
import PgConnectionFilterPlugin from 'postgraphile-plugin-connection-filter'

export async function makeQueryRunner(
  connectionString = process.env.DATABASE_URL || 'postgres:///',
  schemaName = process.env.DATABASE_SCHEMA || 'app_public',
  options = {
    appendPlugins: [PgSimplifyInflectorPlugin, PgConnectionFilterPlugin],
    graphileBuildOptions: {
      connectionFilterComputedColumns: false,
      connectionFilterRelations: true,
      nestedMutationsSimpleFieldNames: true,
    },
  } // See https://www.graphile.org/postgraphile/usage-schema/ for options
) {
  // Create the PostGraphile schema
  const schema = await createPostGraphileSchema(
    connectionString,
    schemaName,
    options
  )

  const pgPool = new Pool({
    connectionString,
  })

  // @ts-ignore
  const apolloClient: ApolloClient<any> = await withPostGraphileContext(
    { pgPool },
    // @ts-ignore
    (context) => {
      const client = new ApolloClient({
        ssrMode: true,
        cache: new InMemoryCache(),
        link: new SchemaLink({ schema, context }),
      })
      return Promise.resolve(client)
    }
  )

  const release = () => {
    pgPool.end()
  }

  return {
    schema,
    release,
    apolloClient,
  }
}
