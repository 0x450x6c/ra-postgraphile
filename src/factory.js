// @flow

import buildGraphQLProvider from 'ra-data-graphql'
import { defaultQueryValueToInputValueMap } from './defaultValueInputTypeMapping'
import type { ProviderOptions } from './types'
import { buildQuery } from './buildQuery'

export const factory = (client: Object, options: ProviderOptions = {}) => {
  const defaultAppliedOptions = {
    queryValueToInputValueMap: {
      ...defaultQueryValueToInputValueMap,
      ...(options.queryValueToInputValueMap || {})
    }
  }

  return buildGraphQLProvider({
    client,
    buildQuery,
    options: defaultAppliedOptions
  })
}
