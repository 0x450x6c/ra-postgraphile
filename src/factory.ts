import buildGraphQLProvider from 'ra-data-graphql'
import type { ApolloClient } from 'apollo-client'
import type { LegacyDataProvider } from 'ra-core'
import { buildQuery } from './buildQuery'
import { defaultQueryValueToInputValueMap } from './defaultValueInputTypeMapping'
import { ProviderOptions, GraphqlProviderOptions } from './types'

export const factory = <T = any>(
  client: ApolloClient<T>,
  options: ProviderOptions = { queryValueToInputValueMap: {} },
  graphqlProviderOptions: GraphqlProviderOptions = {}
): LegacyDataProvider => {
  const defaultAppliedOptions = {
    queryValueToInputValueMap: {
      ...defaultQueryValueToInputValueMap,
      ...(options.queryValueToInputValueMap || {}),
    },
  }

  return buildGraphQLProvider({
    ...graphqlProviderOptions,
    client,
    buildQuery,
    options: defaultAppliedOptions,
  })
}
