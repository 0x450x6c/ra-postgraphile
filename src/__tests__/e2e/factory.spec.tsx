import expect from 'expect'
import { ApolloClient } from 'apollo-client'
import { introspectionFromSchema, IntrospectionQuery } from 'graphql'
import { convertLegacyDataProvider } from 'ra-core'
import { makeQueryRunner } from '../../__test_utils/QueryRunner'
import { factory } from '../../factory'

let client: ApolloClient<any>
let cleanup: () => void
let introspection: IntrospectionQuery

beforeAll(async () => {
  const { schema, release, apolloClient } = await makeQueryRunner()
  cleanup = release
  client = apolloClient
  introspection = introspectionFromSchema(schema, { descriptions: true })
})

afterAll(() => {
  cleanup()
})

describe('factory', () => {
  it('should construct from a schema', async () => {
    const legacyProvider = await factory(
      client,
      { queryValueToInputValueMap: {} },
      { introspection: introspection.__schema }
    )
    const dataProvider = convertLegacyDataProvider(legacyProvider)

    const result = await dataProvider.getList('books', {
      pagination: { page: 1, perPage: 10 },
      filter: {},
      sort: { field: 'id', order: 'ASC' },
    })
    expect(result.data).toMatchSnapshot()
  })
})
