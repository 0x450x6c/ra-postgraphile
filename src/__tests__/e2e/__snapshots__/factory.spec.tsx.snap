// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`factory should construct from a schema 1`] = `
Array [
  Object {
    "__typename": "Book",
    "id": 1,
    "isbn": "12345",
    "name": "Moby Dick or The Whale",
    "nodeId": "WyJib29rcyIsMV0=",
  },
]
`;
